x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/AutoConfigInitialNode.java

```diff
@@ -0,0 +1,352 @@
+                return;
+            }
+            // Silently skipping security auto configuration because the node is configured for cluster formation, and auto-configuration
+            // relies on the node forming a cluster by itself.
+            if (env.settings().hasValue(ClusterBootstrapService.INITIAL_MASTER_NODES_SETTING.getKey())) {
+                terminal.println(Terminal.Verbosity.VERBOSE,
+                        "Skipping security auto configuration because this node is explicitly configured to form a new cluster.");
+                terminal.println(Terminal.Verbosity.VERBOSE,
+                        "The node cannot be auto configured to participate in forming a new multi-node secure cluster.");
+                return;
+            }
+            // Silently skipping security auto configuration because node cannot become master.
+            final List<DiscoveryNodeRole> nodeRoles = NodeRoleSettings.NODE_ROLES_SETTING.get(env.settings());
+            boolean canBecomeMaster = nodeRoles.contains(DiscoveryNodeRole.MASTER_ROLE) &&
+                    false == nodeRoles.contains(DiscoveryNodeRole.VOTING_ONLY_NODE_ROLE);
+            if (false == canBecomeMaster) {
+                terminal.println(Terminal.Verbosity.VERBOSE,
+                        "Skipping security auto configuration because the node is configured such that it cannot become master.");
+                return;
+            }
+            // Silently skipping security auto configuration, because the node cannot contain the Security index data
+            boolean canHoldSecurityIndex = nodeRoles.stream().anyMatch(DiscoveryNodeRole::canContainData);
+            if (false == canHoldSecurityIndex) {
+                terminal.println(Terminal.Verbosity.VERBOSE,
+                        "Skipping security auto configuration because the node is configured such that it cannot contain data.");
+                return;
+            }
+            // Silently skipping security auto configuration because TLS is already configured
+            if (false == env.settings().getByPrefix(XPackSettings.TRANSPORT_SSL_PREFIX).isEmpty() ||
+                    false == env.settings().getByPrefix(XPackSettings.HTTP_SSL_PREFIX).isEmpty()) {
+                // zero validation for the TLS settings as well, let the node bootup do its thing
+                terminal.println(Terminal.Verbosity.VERBOSE,
+                        "Skipping security auto configuration because it appears that TLS is already configured.");
+                return;
+            }
+            // Silently skipping security auto configuration because file realm is configured
+            // Auto-configuration only requires that the file realm be enabled, and, for optimal experience,
+            // also be the first in the chain.
+            if (false == env.settings().getByPrefix(RealmSettings.realmSettingPrefix(FileRealmSettings.TYPE)).isEmpty()) {
```

Comment:
As discussed I think we don't need to care about the file realm. We expect it to be enabled( it is enabled by default in 8.0.0 ) and it is not implicitly disabled in the presence of other realms.
Arguably we could check if the file realm is explicitly disabled as this is an indication of an existing configuration or a case where the security configuration is explicit on behalf of the user and we might not want to attempt to further configure it. But the CLI tools themselves will check that the file realm is enabled and notify the user otherwise, so I think we don't have to check for the file realm here.

Reply:
Don't you think it would be better to skip auto configuration if the file realm is disabled or reordered?
The tools detect it, but that's after auto-configuration did its thing.

Reply:
If the file realm is explicitly disabled, we can take this as a hint that the user wants to own the security related configuration and bail. I'm not convinced it is necessary but it probably fits in our general narrative around "we won't try to auto-configure security if you indicate that you want to handle it yourself". Our auto-configuration process and our enrollment process doen't depend on the CLI tools, it is augmented by it. A user can very well setup a single-node cluster with Kibana , without touching the CLI tools and given what we know of usage for our stack, this is not a small use case. 

But, why does the order of the file realm affect anything? Can you elaborate?

Reply:
> Our auto-configuration process and our enrollment process doen't depend on the CLI tools, it is augmented by it

I'm not sure I would agree that the cli tools augment the enrollment experience. I would rather say that they are needed, otherwise you can't add nodes to the cluster after the first enrollment token expires.

> But, why does the order of the file realm affect anything? Can you elaborate?

Suppose the node is configured with LDAP realm authn. It doesn't look great if our own cmd line tools reach out to the customer's LDAP realms, because the tool iterates the chain and the file realm is the last in the chain.

Reply:
> I'm not sure I would agree that the cli tools augment the enrollment experience. I would rather say that they are needed, otherwise you can't add nodes to the cluster after the first enrollment token expires. 

There is a very common and valid use case that doesn't require the use of the tools. 

I'd be +1 to do this because we think it is an indication of an explicit configuration that we shouldn't try to extend but simply bail, but not because the tools will not be usable. If we do t for the former, maybe we should also bail if _any_ authentication realm has been explicitly configured ? 

> Suppose the node is configured with LDAP realm authn. It doesn't look great if our own cmd line tools reach out to the customer's LDAP realms, because the tool iterates the chain and the file realm is the last in the chain.

Continuing the argument above, if there is an LDAP realm preconfigured on startup, maybe we should not try to configure security at all. 
If this is a later point in the cluster lifecycle when the user has made amendments to the auto-generated configuration and still needs to use the CLI tools, then they should be fine with the the way we do realm chaining in Elasticsearch as this is how the product already works. The CLI tools do nothing extraordinary in that respect.